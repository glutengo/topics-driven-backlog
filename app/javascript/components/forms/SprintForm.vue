<template>
    <b-form @submit.prevent="submit">
        <b-form-group label="Name" label-for="sprint-name">
            <b-form-input id="sprint-name" v-model="data.name" required></b-form-input>
        </b-form-group>

        <b-form-group label="Start date" label-for="sprint-start-date">
            <b-form-input id="sprint-start-date" :max="data.end_date" type="date" v-model="data.start_date" required></b-form-input>
        </b-form-group>

        <b-form-group label="End date" label-for="sprint-end-date">
            <b-form-input id="sprint-end-date" type="date" v-model="data.end_date" required
                          :min="$store.state.currentDate > data.start_date ? $store.state.currentDate : data.start_date"></b-form-input>
        </b-form-group>

        <b-button type="submit" variant="primary" :disabled="isInvalid">Save</b-button>
        <b-button type="button" variant="grey" @click="cancel">Cancel</b-button>
    </b-form>
</template>

<script>
    import {required} from 'vuelidate/lib/validators';
    import smallerOrEqualThan from '../../validators/smallerOrEqualThan';
    import largerOrEqualThan from '../../validators/largerOrEqualThan';
    import BaseForm from './BaseForm.vue';
    import moment from "moment";

    export default {
        name: 'sprint-form',
        extends: BaseForm,
        validations: {
            data: {
                name: {required},
                start_date: {
                    required,
                    smallerThanEnd: smallerOrEqualThan('end_date')
                },
                end_date: {
                    required,
                    largerThanStart: largerOrEqualThan('start_date'),
                    largerThanToday: largerOrEqualThan(moment().format('YYYY-MM-DD'), true),
                },
            },
        },
    }
</script>