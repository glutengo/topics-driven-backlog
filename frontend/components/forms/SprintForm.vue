<template>
    <BForm 
        @submit.prevent="submit" 
        class="sprint-form">

        <BFormGroup 
            label="Name" 
            label-for="sprint-name">

            <BFormInput 
                id="sprint-name" 
                v-model="data.name" 
                required/>
        </BFormGroup>

        <BFormGroup 
            label="Start date" 
            label-for="sprint-start-date">

            <BFormInput 
                id="sprint-start-date" 
                :max="data.end_date" 
                type="date" 
                v-model="data.start_date" 
                required/>
        </BFormGroup>

        <BFormGroup 
            label="End date" 
            label-for="sprint-end-date">

            <BFormInput 
                id="sprint-end-date" 
                type="date" 
                v-model="data.end_date" 
                required
                :min="$store.state.currentDate > data.start_date ? $store.state.currentDate : data.start_date"/>
        </BFormGroup>

        <BButton 
            type="submit" 
            variant="primary" 
            :disabled="isInvalid">Save</BButton>

        <BButton 
            type="button" 
            variant="grey" 
            @click="cancel">Cancel</BButton>
    </BForm>
</template>

<script>
import moment from 'moment';
import { required } from 'vuelidate/lib/validators';
import BForm from '@bootstrap/form/form';
import BFormGroup from '@bootstrap/form-group/form-group';
import BFormInput from '@bootstrap/form-input/form-input';
import BButton from '@bootstrap/button/button';
import smallerOrEqualThan from '@/validators/smallerOrEqualThan';
import largerOrEqualThan from '@/validators/largerOrEqualThan';
import BaseForm from '@/components/forms/BaseForm';

export default {
    name: 'SprintForm',
    components: { BForm, BFormGroup, BFormInput, BButton },
    extends: BaseForm,
    validations: {
        data: {
            name: { required },
            start_date: {
                required,
                smallerThanEnd: smallerOrEqualThan('end_date'),
            },
            end_date: {
                required,
                largerThanStart: largerOrEqualThan('start_date'),
                largerThanToday: largerOrEqualThan(
                    moment().format('YYYY-MM-DD'),
                    true
                ),
            },
        },
    },
};
</script>