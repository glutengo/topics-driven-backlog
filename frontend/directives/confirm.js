import Vue from 'vue';

/**
 * v-confirm directive, asking the user to confirm the action before
 * a click event on an item is processed
 */
Vue.directive('confirm', {
    bind(el, binding) {
        el.handleClick = () => {
            if (confirm('Are you sure? This action is irreversible.')) {
                binding.value();
            }
        };

        el.addEventListener('click', el.handleClick);
    },
    unbind(el) {
        el.removeEventListener('click', el.handleClick);
    },
});
