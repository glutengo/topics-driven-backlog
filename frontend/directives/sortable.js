import Vue from 'vue';
import * as Sortable from 'sortablejs/Sortable';

/**
 * v-sortable directive, binding Sortable plugin to a list
 */
Vue.directive('sortable', {
    inserted: function(el, binding, vnode) {
        if (binding.value) {
            vnode.context.sortableInstance = Sortable.create(el, binding.value);
        }
    },
});
