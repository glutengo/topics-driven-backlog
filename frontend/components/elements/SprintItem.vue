<template>
    <li class="sprint-item card">
        <div class="card-body">
            <template v-if="!editing">
                <template v-if="data.end_date >= currentDate">
                    <BButton 
                        size="sm" 
                        variant="danger" 
                        class="float-right ml-2" 
                        v-confirm="deleteSprint">Delete</BButton>

                    <BButton 
                        size="sm" 
                        variant="primary" 
                        class="float-right" 
                        @click="startEditing">Edit</BButton>
                </template>
                
                <h3 class="card-title h5">{{ data.name }}</h3>

                <p class="card-text">
                    Start: {{ data.start_date | displayDate }} <br>
                    End: {{ data.end_date | displayDate }}
                </p>
            </template>

            <SprintForm 
                v-else 
                v-model="editingData" 
                @cancel="editing = false" 
                @submit="saveSprint"/>
        </div>
    </li>
</template>

<script>
import * as _ from 'lodash';
import moment from 'moment';
import BButton from '@bootstrap/button/button';
import SprintForm from '@/components/forms/SprintForm';

export default {
    name: 'SprintItem',
    components: { SprintForm, BButton },
    props: {
        data: {
            type: Object,
            default: () => {},
        },
    },
    data() {
        return {
            editing: false,
            editingData: null,
            currentDate: moment().format('YYYY-MM-DD'),
        };
    },
    methods: {
        /**
         * Start editing process of this sprint
         */
        startEditing() {
            this.editing = true;
            this.editingData = _.pick(this.data, [
                'name',
                'start_date',
                'end_date',
            ]);
        },

        /**
         * Save the edited parameters of this sprint
         */
        async saveSprint() {
            await this.$store.dispatch('sprints/update', {
                id: this.data.id,
                parentId: this.data.course_id,
                ...this.editingData,
            });

            this.editing = false;

            // TODO handle errors in UI
        },

        /**
         * Delete this sprint
         */
        async deleteSprint() {
            await this.$store.dispatch('sprints/remove', {
                id: this.data.id,
                parentId: this.data.course_id,
            });

            // TODO handle errors in UI
        },
    },
};
</script>