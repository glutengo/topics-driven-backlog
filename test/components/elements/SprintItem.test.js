import { mount, createLocalVue } from 'vue-test-utils';
import moment from 'moment';
import Vuex from 'vuex';
import Vuelidate from 'vuelidate';
import SprintItem from '@/components/elements/SprintItem';
import SprintForm from '@/components/forms/SprintForm';
import '@/directives/confirm';

// take a date far in the future so validations pass
Date.now = jest.fn(() => 7956915742000);

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('SprintItem.test.js', () => {
    let cmp, actions, store;

    beforeEach(() => {
        actions = {
            patch: jest.fn(),
            delete: jest.fn(),
        };

        store = new Vuex.Store({
            modules: {
                sprints: {
                    namespaced: true,
                    actions,
                },
            },
        });

        cmp = mount(SprintItem, {
            localVue,
            store,
            propsData: {
                data: {
                    id: 1,
                    name: 'Test Sprint',
                    start_date: moment()
                        .add(5, 'days')
                        .format('YYYY-MM-DD'),
                    end_date: moment()
                        .add(7, 'days')
                        .format('YYYY-MM-DD'),
                },
            },
        });
    });

    it('shows edit and delete buttons when end_date not in the past', () => {
        expect(cmp.findAll('.card-body .btn').length).toBe(2);
        expect(cmp.element).toMatchSnapshot();
    });

    it('calls starts editing when edit button is clicked', () => {
        cmp.vm.startEditing = jest.fn();
        cmp.update();

        cmp
            .findAll('.card-body .btn')
            .at(1)
            .trigger('click');

        expect(cmp.vm.startEditing).toBeCalled();
    });

    it('is in edit mode after startEditing was called', () => {
        cmp.vm.startEditing();
        cmp.update();
        expect(cmp.vm.editing).toBe(true);
        expect(cmp.vm.editingData).toBeTruthy();
        expect(cmp.element).toMatchSnapshot();
    });

    it('triggers sprint save when sprint form triggers submit', () => {
        cmp.vm.startEditing();
        cmp.vm.saveSprint = jest.fn();
        cmp.update();        

        cmp.find(SprintForm).trigger('submit');
        expect(cmp.vm.saveSprint).toBeCalled();
    });

    it('patches the sprint and leaves editing mode on sprint save', () => {
        cmp.vm.saveSprint();

        expect(actions.patch).toBeCalled();
        expect(cmp.vm.editing).toBe(false);
    });

    it('deletes the sprint when delete button is clicked', () => {
        const confirmSpy = jest
            .spyOn(window, 'confirm')
            .mockImplementation(() => true);

        cmp
            .findAll('.card-body .btn')
            .at(0)
            .trigger('click');

        expect(confirmSpy).toHaveBeenCalled();
        expect(actions.delete).toHaveBeenCalled();
    });
});
