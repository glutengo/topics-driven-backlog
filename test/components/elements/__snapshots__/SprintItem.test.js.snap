// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprintItem.test.js is in edit mode after startEditing was called 1`] = `
<li
  class="sprint-item card"
>
  <div
    class="card-body"
  >
    <form
      class="sprint-form"
    >
      <div
        aria-labelledby="__BVID__22___BV_label_"
        class="b-form-group form-group"
        id="__BVID__22_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="sprint-name"
          id="__BVID__22___BV_label_"
        >
          Name
        </label>
        <div
          class=""
        >
          <input
            aria-required="true"
            class="form-control"
            id="sprint-name"
            required="required"
            type="text"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
       
      <div
        aria-labelledby="__BVID__24___BV_label_"
        class="b-form-group form-group"
        id="__BVID__24_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="sprint-start-date"
          id="__BVID__24___BV_label_"
        >
          Start date
        </label>
        <div
          class=""
        >
          <input
            aria-required="true"
            class="form-control"
            id="sprint-start-date"
            max="2018-01-03"
            required="required"
            type="date"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
       
      <div
        aria-labelledby="__BVID__26___BV_label_"
        class="b-form-group form-group"
        id="__BVID__26_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="sprint-end-date"
          id="__BVID__26___BV_label_"
        >
          End date
        </label>
        <div
          class=""
        >
          <input
            aria-required="true"
            class="form-control"
            id="sprint-end-date"
            min="2018-01-01"
            required="required"
            type="date"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
       
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save
      </button>
       
      <button
        class="btn btn-grey"
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</li>
`;

exports[`SprintItem.test.js shows edit and delete buttons when end_date not in the past 1`] = `
<li
  class="sprint-item card"
>
  <div
    class="card-body"
  >
    <button
      class="btn float-right ml-2 btn-danger btn-sm"
      type="button"
    >
      Delete
    </button>
     
    <button
      class="btn float-right btn-primary btn-sm"
      type="button"
    >
      Edit
    </button>
     
    <h3
      class="card-title h5"
    >
      Test Sprint
    </h3>
     
    <p
      class="card-text"
    >
      
                Start: 2018-01-01 
      <br />
      
                End: 2018-01-03
            
    </p>
  </div>
</li>
`;
